cpp/monitoring/gcm/exporter.cc

```diff
@@ -0,0 +1,264 @@
+#include "monitoring/gcm/exporter.h"
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+#include <sstream>
+
+#include "monitoring/metric.h"
+#include "monitoring/registry.h"
+#include "net/url.h"
+#include "util/json_wrapper.h"
+#include "util/sync_task.h"
+
+DEFINE_string(google_compute_monitoring_push_url, "",
+              "GCM custom timeseries write URL.");
+DEFINE_string(google_compute_monitoring_metrics_url, "",
+              "GCM custom timeseries metrics URL.");
+DEFINE_int32(google_compute_monitoring_push_interval_seconds, 5,
+             "Seconds between pushing metric values to GCM.");
+DEFINE_string(google_compute_metadata_url,
+              "http://metadata/computeMetadata/v1/instance/service-accounts",
+              "URL of GCE metadata server.");
+DEFINE_string(google_compute_monitoring_service_account, "default",
+              "Which GCE service account to use for pushing metrics.");
+
+namespace cert_trans {
+
+using std::chrono::seconds;
+using std::chrono::system_clock;
+using std::chrono::minutes;
+using std::make_pair;
+using std::mutex;
+using std::ostringstream;
+using std::placeholders::_1;
+using std::string;
+using std::thread;
+using std::unique_lock;
+using util::Executor;
+using util::SyncTask;
+
+
+namespace {
+
+
+// Label and Metric -name prefix.
+const char kCloudPrefix[] = "custom.cloudmonitoring.googleapis.com/ct/";
+
+
+inline std::string RFC3339Time(const system_clock::time_point& when) {
+  const std::time_t now_c(system_clock::to_time_t(when));
+  char buf[256];
+  CHECK(std::strftime(buf, sizeof(buf), "%FT%T.00Z", std::localtime(&now_c)));
+  return buf;
+}
+
+
+}  // namespace
+
+
+GCMExporter::GCMExporter(const string& instance_name, UrlFetcher* fetcher,
```

Comment:
My point is that the convention of passing (Foo* foo) generally denotes an ownership transfer, which doesn't occur in this case.